src/main/java/edu/wpi/grip/core/sources/WebCamSource.java

```
@@ -0,0 +1,131 @@
+    }
+
+    @Override
+    public OutputSocket[] getOutputSockets() {
+        return new OutputSocket[]{frameOutputSocket, frameRateOutputSocket};
+    }
+
+    /**
+     * Starts the video capture from the
+     * @param deviceNumber The type of the device the webcam should be attached with
+     */
+    public synchronized void startVideo(final int deviceNumber) {
+        final OpenCVFrameConverter.ToMat convertToMat = new OpenCVFrameConverter.ToMat();
+        final OpenCVFrameGrabber grabber = new OpenCVFrameGrabber(deviceNumber);
+        if(this.executor.isPresent()){
+            throw new IllegalStateException("The video retrieval thread has already been started.");
+        }
+        if(this.grabber.isPresent()){
+            throw new IllegalStateException("The Frame Grabber has already been started.");
+        }
+        try {
+            grabber.start();
+        } catch (FrameGrabber.Exception e) {
+            throw new IllegalStateException("A problem occurred trying to start the frame grabber", e);
+        }
+        // Store the grabber only once it has been started in the case that there is an exception.
+        this.grabber = Optional.of(grabber);
+
+        final Thread frameExecutor = new Thread(() -> {
+            long lastFrame = System.currentTimeMillis();
+            while(!Thread.interrupted()) {
+                final Frame videoFrame;
+                try {
+                    videoFrame = grabber.grab();
+                } catch (FrameGrabber.Exception e) {
+                    throw new IllegalStateException("Failed to grab image", e);
+                }
+                frameOutputSocket.setValue(convertToMat.convert(videoFrame));
+                long thisMoment = System.currentTimeMillis();
+                frameRateOutputSocket.setValue(1000/(thisMoment - lastFrame));
+                lastFrame = thisMoment;
+            }
+        });
+        frameExecutor.setUncaughtExceptionHandler(
```

Since the entire thread is defined above, it would probably be better to catch all exceptions within the thread.